edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/CloudConnectionProvider.cs

```diff
@@ -69,21 +68,119 @@ public void BindEdgeHub(IEdgeHub edgeHubInstance)
+                    this.clientProvider,
+                    cloudListener,
+                    tokenProvider,
+                    this.idleTimeout,
+                    this.closeOnIdleTimeout,
+                    this.operationTimeout,
+                    productInfo,
+                    modelId);
+            }
+            else
+            {
+                throw new AuthenticationException($"Unabled to connectwith credentials type {deviceCredentials.GetType()} for device {identity}");
+            }
+
+            Events.SuccessCreatingCloudConnection(identity);
+            return cloudConnection;
+        }
+
+        public Task<ITry<ICloudConnection>> Connect(IClientCredentials clientCredentials, Action<string, CloudConnectionStatus> connectionStatusChangedHandler)
         {
             Preconditions.CheckNotNull(clientCredentials, nameof(clientCredentials));
+            return Option.Some(clientCredentials)
+                .Map(cr => cr as ITokenCredentials)
+                .Map(tcr => this.ConnectAsync(tcr.Identity, connectionStatusChangedHandler, Option.Some(clientCredentials)))
+                .GetOrElse(() => this.Connect(clientCredentials.Identity, connectionStatusChangedHandler));
+        }
 
+        public Task<ITry<ICloudConnection>> Connect(IIdentity identity, Action<string, CloudConnectionStatus> connectionStatusChangedHandler) => this.ConnectAsync(identity, connectionStatusChangedHandler, Option.None<IClientCredentials>());
+
+        async Task<ITry<ICloudConnection>> ConnectAsync(IIdentity identity, Action<string, CloudConnectionStatus> connectionStatusChangedHandler, Option<IClientCredentials> initialCredentials)
+        {
+            Preconditions.CheckNotNull(identity, nameof(identity));
             try
             {
-                Events.CreatingCloudConnectionUsingClientCredentials(clientCredentials);
-                var cloudListener = new CloudListener(this.edgeHub.Expect(() => new InvalidOperationException("EdgeHub reference should not be null")), clientCredentials.Identity.Id);
-                ConnectionMetadata connectionMetadata = await this.metadataStore.GetMetadata(clientCredentials.Identity.Id);
+                ConnectionMetadata connectionMetadata = await this.metadataStore.GetMetadata(identity.Id);
                 string productInfo = connectionMetadata.EdgeProductInfo;
-                Option<string> modelId = clientCredentials.ModelId.HasValue ? clientCredentials.ModelId : connectionMetadata.ModelId;
-                if (this.edgeHubIdentity.Id.Equals(clientCredentials.Identity.Id))
+                Option<string> modelId = connectionMetadata.ModelId;
+
+                var serviceIdentity = await this.deviceScopeIdentitiesCache.GetServiceIdentity(identity.Id, true);
+                var cloudConnectionCreationTask = serviceIdentity.Map(si => this.ConnectOnBehalfOfAsync(si, identity, connectionStatusChangedHandler, productInfo, modelId))
```

Comment:
>serviceIdentity.Map [](start = 50, length = 19)

IIRC there's a special-case here where the EdgeHub connects for itself before the identity cache finishes the initial sync.  It might still be okay as we'd have the module token, but it'd go down the ConnectWithoutScope path and seems a little weird.

Reply:
Isn't EdgeHub put into identity cache as initial setup? EHC even couldn't startup without it.